snap/snapcraft.yaml

```diff
@@ -0,0 +1,35 @@
+name: webthings-gateway
+version: 0.10.0
+summary: WebThings Gateway by Mozilla
+description: Directly monitor and control your home over the web.
+
+base: core18
+grade: devel
+confinement: devmode
+
+parts:
+  webthings-gateway:
+    source: https://github.com/mozilla-iot/gateway.git
+    source-type: git
+    source-tag: 0.10.0
+    plugin: better-nodejs
+    nodejs-package-manager: npm
+    nodejs-version: 8.17.0
+    build-packages:
+      - build-essential
+      - git
+      - pkg-config
+      - python
+    override-build: |
+      export PATH=$PATH:$(pwd)/../npm/bin
+      npm config -g set unsafe-perm true
+      npm install -g node-gyp
+      rm -rf ./node_modules package-lock.json
+      (
+        set -e
+        unset SUDO_UID
+        unset SUDO_GID
```

Comment:
Am I right in thinking this is to make `npm install` run as root?

I tried a variation on this myself but couldn't get it to work. I'd be interested in a line by line explanation of what these 7 lines are doing, including why the subshell is needed. I would like to learn from your bash foo!

Reply:
npm was already running as root via the nodejs plugin. The override-build stage here is essentially replicating what the plugin was doing, but adding in webpack and taking out the weird offline caching bit.

```sh
# Add the locally installed npm to our path.
# export makes it also propagate into the sub-shell.
export PATH=$PATH:$(pwd)/../npm/bin

# Allow npm to run as root properly
# https://docs.npmjs.com/misc/config#unsafe-perm
npm config -g set unsafe-perm true

# Install node-gyp, as it didn't seem to get installed
# by `npm install` for some reason
npm install -g node-gyp

# Clean up and let npm do a clean install. I ended up
# having to do this for all the other packages as well.
rm -rf ./node_modules package-lock.json

# Start a sub-shell so that the SUDO_* bits aren't
# propagated to the rest of the process unnecessarily.
(
  # Make the subshell exit if one of the commands fails
  set -e

  # Unset the sudo environment variables to make git happy
  # https://github.com/snapcore/snapcraft/commit/93fd2825733ea8725502d2ab62124be5179c6407
  unset SUDO_UID
  unset SUDO_GID

  # Install all of the necessary node modules
  npm install

  # Run the webpack process now, rather than at each launch
  ./node_modules/.bin/webpack

  # Remove all but production dependencies
  npm prune --production
)
```

Reply:
Thanks for this great explanation. If it's possible to put code comments in a `snapcraft.yaml` it might be useful to add these comments to the file for others...

Reply:
Done -- I pushed a commit with comments.